---
title: "SPRUCE - Single-cell Pairwise Relationships Untangled by Composite ETM"
author: "Park Lab"
date: "`r format(Sys.time(), ' %X, %b %d, %Y')`"
fontsize: 11pt
output:
  pdf_document:
    latex_engine: xelatex
    citation_package: natbib
    fig_caption: true
bibliography: references.bib 
link-citations: yes
tables: yes
always_allow_html: true
header-includes:
 - \usepackage{fontspec}
 - \usepackage{booktabs}
 - \usepackage{longtable}
 - \usepackage{array}
 - \usepackage{multirow}
 - \usepackage{wrapfig}
 - \usepackage{float}
 - \usepackage{colortbl}
 - \usepackage{pdflscape}
 - \usepackage{tabu}
 - \usepackage{threeparttable}
 - \usepackage{threeparttablex}
 - \usepackage{makecell}
 - \usepackage{amsmath}
 - \usepackage{amssymb}
 - \usepackage{setspace}\doublespacing
---

## Methods

### Data origins and preprocessing
We generated a mixuture dataset combining three different recent breast cancer studies. The first one is a breast cancer dataset consisting of 100k cells from a single-cell atlas of human breast cancers $\cite{wu2021single}. The secod dataset consited of 48k cells from a single-cell atlas of the healthy breast tissues (\cite{bhat2021single}).The third dataset is 6k subset of breast cancer CD4 and CD8 T-cells from a pan-cancer atlas of tumour-infiltrating T cells profiled across 21 cancer types and 316 donors (\cite{zheng2021pan}). The total number of cells in the combined dataset was 155913. We filtered out genes detected in less than 3 cells along with mitochondrial gene and spike genes, which lead to 20265 genes in the final dataset. 

### The SPRUCE Model
The SPRUCE model consists of two steps of topic modeling - first, we model cell topic to assign each cell into type/state, and then we use these cell topics to construct a set of neighbours and model interaction topic, which assigns each neighbour pair to unique interaction state. 

### Cell topic analysis
The cell topic modeling extends on the ideas of LDA. Consider a sample of cells ${i_{1},....,i_{N}}$ and a list of genes ${g_{1},....,g_{G}}$, where ${x_{i1},x_{i2},...,x_{iG}}$ are raw count data for $G$ genes in cell $i$. We assume that each cell count vector $X_{i}$ was generated from a multinomial distribution parameterized by a gene expression frequency matrix with an element $\rho_{ig}$ of a gene $g$ in the cell $i$.

$$p(\mathbf{x}_{i}|\rho_{i}) = \prod_{g} \rho_{ig}^{X_{ig}}$$

We introduce Dirichlet prior on the $\rho$ and parameterize the Dirichlet as a generalized linear model (GLM). Exploiting the conjugacy between the multinomial and Dirichlet, we integrate out the unknown parameters $\rho$.

$$p(\rho_{i}|\lambda_{i}) = \frac{\Gamma(\sum_{g} \lambda_{ig})}{\prod_{g} \Gamma(\lambda_{ig})} \prod_{g \in \textsf{genes}} \rho_{ig}^{\lambda_{ig} - 1}$$

$$\lambda_{ig} = \exp\left( \sum_{t=1}^{T} \theta_{it} ( \beta_{tg} + b_{g} ) \right)$$

The topic proportion $\theta_{it}$ for cell $i$ is drawn from logistic normal distribution with model hyperparameters $\delta_{it}$ and we assume topic proportions within a simplex, namely $\sum_{t\in \textsf{topics}} \theta_{it} = 1$.

\begin{equation}
\begin{split}
\delta_{it} \sim N(0,I);\theta_{it} = softmax(\delta_{it})\\
\end{split}
\end{equation}

The marginal likelihood 

$$p(\mathbf{x}_{i}| \cdot ) = 
\frac{\Gamma(\sum_{g} \lambda_{ig})} {\sum_{g} \Gamma(\lambda_{ig})} \frac{\Gamma(\sum_{g} \lambda_{ig} + X_{ig})}{\sum_{g} \Gamma(\lambda_{ig} + X_{ig})}$$


The marginal likelihood of each cell is an intractable problem because it involves integral over the topic proportion. Variational inference techniques can be used to approximate this type of intractable integrals. 

Let $p_{\theta}(z \mid x)$ be a true posterior and $q_{\phi}(z \mid x)$ be an approximate posterior.

\begin{equation}
\begin{split}
D_{KL}(q_{\phi}\mid\mid p_{\theta}) & = E_{q_{\phi}}[log  \frac {q_{\phi}(z \mid x)}{p_{\theta}(z \mid x)}]\\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z \mid x)}]\\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ \frac {p_{\theta}(z,x)}{p_{\theta}(x)}]\\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] + E_{q_{\phi}} [log \ {p_{\theta}(x)}]\\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] + \int q_{\phi}(z \mid x) log \ {p_{\theta}(x)}dz\\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] + log \ {p_{\theta}(x)} \int q_{\phi}(z \mid x) dz \\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] + log \ {p_{\theta}(x)}\\
log \ {p_{\theta}(x)} & = - E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] + E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] + D_{KL}(q_{\phi}\mid\mid p_{\theta}) \\
\end{split}
\end{equation}

Here, $D_{KL}(q_{\phi}\mid\mid p_{\theta})$ is intractable but it is always $\geq$ 0, we can use this property to remove $D_{KL}$ from the equation and the marginal log likelihood $log \ {p_{\theta}(x)}$ will be at least $\geq - E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] + E_{q_{\phi}} [log \ {p_{\theta}(z,x)}]$. Since this term is the lower bound on the evidence,it is called as Evidence Lower Bound(ELBO). We maximize the marginal log likelihood by maximizing the ELBO and indirectly minimize the KL divergence.

\begin{equation}
\begin{split}
ELBO & = - E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] + E_{q_{\phi}} [log \ {p_{\theta}(z,x)}] \\
& = - E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] + E_{q_{\phi}} [log \ {p_{\theta}(x \mid z)}] + E_{q_{\phi}} [log \ {p_{\theta}(z)}] \\
& = E_{q_{\phi}} [log \ {p_{\theta}(x \mid z)}]  - E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] + E_{q_{\phi}}[log \ {p_{\theta}(z)}] \\
ELBO & = E_{q_{\phi}} [log \ {p_{\theta}(x \mid z)}]  - E_{q_{\phi}} [log \frac{q_{\phi}(z \mid x)}{p_{\theta}(z)}] \\
\end{split}
\end{equation}

Here, $E_{q_{\phi}} [log \ {p_{\theta}(x \mid z)}]$ is an expected reconstruction error and $E_{q_{\phi}} [log \frac{q_{\phi}(z \mid x)}{p_{\theta}(z)}]$ is KL Divergence between approximate posterior and the prior.


Approximate posterior and prior

The approximate posterior $q_{\phi}(z \mid x)$ is a Gaussian variational distribution $q(\delta_{i};i_{n},v)$ = $N(\mu,\Sigma)$ whose mean and variance are constructed form a neural network parameterized by $v$. The network takes raw count data of a cell 
$i_{n}$ for $G$ genes and outputs a mean and variance of $\delta_{i}$. The prior $p_{\theta}(z)$ = $N(0,I)$. The KL divergence between these two form of Gaussians exist in closed form and given as-

\begin{equation}
\begin{split}
D_{KL}(q_{\phi}\mid\mid p_{\theta}) & = E_{q_{\phi}} [log \frac{q_{\phi}(z \mid x)}{p_{\theta}(z)}] \\
& = E_{q_{\phi}}[log \ {q_{\phi}(z \mid x)}] - E_{q_{\phi}} [log \ {p_{\theta}(z)}]\\
& = 1/2 \sum_{d} (1 + log(\Sigma) - \mu^{2} - \Sigma) \\
\end{split}
\end{equation}

In addition to the latent state KL divergence, we take into account the uncertainty of $\beta_{tg}$ parameters:

$$\beta_{tg} \sim \mathcal{N}\!\left(0, 1\right)$$

Total Expected log-likelihood Lower-bound (ELBO):

\begin{align}
\frac{J}{n} 
    &=\frac{1}{n}\sum_{i=1}^{n} \log p(\mathbf{x}_{i}|\theta_{i}(\mathbf{z}_{i}), \beta) \\
	&  + \frac{1}{n}\sum_{i=1}^{n} \sum_{t=1}^{T} D_{\textsf{KL}}\left( q(z_{it}) \| p(z_{it}) \right) \\
	&  + \frac{1}{n}\sum_{t=1}^{T}\sum_{g=1}^{G} D_{\textsf{KL}}\left( q(\beta_{tg}) \| p(\beta_{tg}) \right) \\
\end{align}

\newpage

### Neighbour cells calculation

A set of neighbour cells were calculated for each cell using the topic assignment from cell topic model. For each cell, five neighbours from each topic were calculated using ANNOY. In total, we generated 159 neighbours for each cell - 32 topics and 5 neighbours from each topic.
- removed self neighbour pair. 18 topics with cell count less than 100 were removed during generating annoy model list. Topics were not removed during generating neighbours, only selected topics were used to create a model list. 

### Ligan receptor data augmentation

For each cell pair, we selected ligand and receptors genes from celltalkDB database and incorporated known interactions from the database using the transformation of spaces as shown in the figure.

  
\newpage 


### Interaction topic analysis


Multinomial-Dirichlet:


$$p(\mathbf{y}_{i}|\mathbf{q}_{i}) = \frac{(\sum_{g} Y_{ig})!}{\prod_{g} Y_{ig}!} \prod_{g} q_{ig}^{Y_{ig}}$$

$$\mathbf{q}_{i} \sim \textsf{Dir}(\mathbf{q}_{i}|\rho_{i}) = \frac{\Gamma(\sum_{g}\rho_{ig})}{\prod_{g} \Gamma(\rho_{ig})} \prod_{g} q_{ig}^{\rho_{ig} - 1}$$


Single-cell generative model:
$$p(\mathbf{x}_{j}| \cdot ) = 
\frac{\Gamma(\sum_{g} \lambda_{jg})} {\sum_{g} \Gamma(\lambda_{jg})} \frac{\Gamma(\sum_{g} \lambda_{jg} + X_{jg})}{\sum_{g} \Gamma(\lambda_{jg} + X_{jg})}$$

where

$$\lambda_{jg} =  \lambda_{0}  \exp\left( \sum_{t=1}^{T} \theta_{jt} ( \beta_{tg} + \delta_{g} ) \right)$$

$$\lambda_{0} = \exp(\tilde{\lambda}_{0})$$


$\sum_{t} \theta_{jt} = 1$ 

Bayesian regularization of the model parameters

$$\beta_{tg} \sim \mathcal{N}\!\left(0, 1\right)$$

Total Expected log-likelihood Lower-bound (ELBO):


\begin{align}
\frac{J}{n} 
    &= \frac{1}{n}\sum_{i=1}^{n} \log p(\mathbf{x}_{i}|\theta_{i}(\mathbf{z}_{i}), \beta) \\
	&  + \frac{1}{n}\sum_{i=1}^{n} \sum_{t=1}^{T} D_{\textsf{KL}}\left( q(z_{it}) \| p(z_{it}) \right) \\
	&  + \frac{1}{n}\sum_{t=1}^{T}\sum_{l=1}^{L} D_{\textsf{KL}}\left( q(\beta_{tl}) \| p(\beta_{tl}) \right) \\
	&  + \frac{1}{n}\sum_{t=1}^{T}\sum_{r=1}^{R} D_{\textsf{KL}}\left( q(\beta_{tr}) \| p(\beta_{tr}) \right) 
\end{align}



# Results

## Probabilistic topic models identify resident cell types and cancer subtypes

## Cell-cell interaction topics reveal new cancer types

## Different interaction topics induce subtype-specific gene-gene networks 

## Each interaction topic show disjoint differential expression genes

- State:24
Ligand S100A9 in cancer and receptor CD68 in neighbouring B cells/myeloid cells.
S100A9 is a calcium-binding protein that is associated with inflammation and expressed not only in myeloid cells but also in some tumours.
S100A9 expressed in ER−PgR− breast cancers induces inflammatory cytokines and is associated with an impaired overall survival.
British Journal of Cancer volume 113, pages1234–1243 (2015)
Ligand HSPA1A in cancer
LncRNA HOTAIR enhances breast cancer radioresistance through facilitating HSPA1A expression via sequestering miR-449b-5p


- State:22
B2M is abnormally expressed in many cancer types.
B2M binds to CD3D,CD3G,KLRD1, etc.
The chemokine receptor CXCR4 has been found to be a prognostic marker in various types of cancer, including breast cancer. 


\newpage

# SUPPLEMENTAL


Total number of cells from different datasets-

```json
GSE164898            48495 

T-cells              35214 
Cancer Epithelial    24489 
Myeloid               9675 
Endothelial           7605 
CAFs                  6573 
PVL                   5423 
Normal Epithelial     4355 
Plasmablasts          3524 
B-cells               3206 --> 101149 

GSE156728-CD4         3063 
GSE156728-CD8         4291 --> 6269
```


Removed cell topics during generating neighbour model list.
```json
h35       57
h11       50
h10       39
h8        36
h44       34
h42       20
h29       18
h3        16
h5        13
h16       12
h36        7
h25        6
h47        6
h41        5
h49        4
h15        3
h18        2
h13        2
```